### PR TITLE
Add Node 16.x to CI

### LINKED ISSUE #1216: Add Node.js 16 to GitHub Action CI builds
### Description

Node.js 16 is now available. 
https://nodejs.medium.com/node-js-16-available-now-7f5099a97e70

We can start testing with the runtime.
https://github.com/slackapi/node-slack-sdk/blob/d4e89b7270c07423656f7a6fea52782345aa9254/.github/workflows/ci-build.yml#L20

### What type of issue is this? (place an `x` in one of the `[ ]`)
- [ ] bug
- [ ] enhancement (feature request)
- [ ] question
- [ ] documentation related
- [x] testing related
- [ ] discussion

### Requirements (place an `x` in each of the `[ ]`)
* [x] I've read and understood the [Contributing guidelines](https://github.com/slackapi/node-slack-sdk/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've searched for any related issues and avoided creating a duplicate issue.

### PR DESCRIPTION
**Pull request by connorads**
###  Summary

Closes #1216

Tried running locally and it failed on linting, want to see if it happens in CI too. If so, might need to work on #1074

Also removed `TODO` about [Node 10 as it is EOL](https://endoflife.date/nodejs)

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/node-slack-sdk/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

### CODE REVIEW

**Comment by seratch** [comment]
Can you check the following tslint error? 

```
> @slack/events-api@3.0.0 lint
> tslint --project .


ERROR: /home/runner/work/node-slack-sdk/node-slack-sdk/packages/events-api/src/adapter.ts:91:13 - Expected parentheses around arrow function argument.
ERROR: /home/runner/work/node-slack-sdk/node-slack-sdk/packages/events-api/src/verify.ts:37:9 - Expected parentheses around arrow function argument.
```

Perhaps, `.then(server => ` needs to be `.then((server) => `, which is a backward-compatible change.

**Comment by connorads** [comment]
Sure @seratch I've fixed the easy ones.
_Although it's a bit weird that `tslint` seems to fail under different Node versions 👀_

There's a bit of a weird one now in
```typescript


```

image

File: packages/events-api/src/adapter.ts [hunk 1]
```diff
@@ -88,7 +88,7 @@ export class SlackEventAdapter extends EventEmitter {
    */
   public start(port: number): Promise<http.Server> {
     return this.createServer()
-      .then(server => new Promise((resolve, reject) => {
```
Comment (connorads) [inline packages/events-api/src/adapter.ts:91]: [Node 12 build is now failing ❌](https://github.com/slackapi/node-slack-sdk/pull/1249/checks?check_run_id=2717943160)
image
The error seems to suggest that there aren't any curly braces when there are ... Seems like `tslint` is struggling to parse the code correctly. I wonder if there's a nice way of resolving this without many ignore comments without replacing `tslint` with `eslint` as per #1074

Comment (connorads) [reply packages/events-api/src/adapter.ts:91]: Actually tell a lie, I can just try sticking curly braces immediately after `server => { return new Promise ...`
image

Comment (seratch) [reply packages/events-api/src/adapter.ts:91]: >I wonder if there's a nice way of resolving this without many ignore comments without replacing tslint with eslint as per #1074

Yeah...  I am thinking so too.

Comment (connorads) [reply packages/events-api/src/adapter.ts:91]: Okay so I think this is sorted now with https://github.com/slackapi/node-slack-sdk/pull/1249/commits/8e37ed7852256d36e6d3017f5a3ea7ac746d298d 🤞

What do you think about the other comment about ignoring the error for `ack` @seratch?

Comment (seratch) [reply packages/events-api/src/adapter.ts:91]: @connorads Sorry, I overlooked it 🤦 

>can we instead just ignore it? What do you think about adding an ignore for this @seratch?

Thanks for checking this in details. Ignoring the lint warning sounds fine 👍

File: packages/socket-mode/src/SocketModeClient.ts [hunk 2]
```diff
@@ -573,6 +573,7 @@ export class SocketModeClient extends EventEmitter {
     }
 
     // Define Ack
+    // tslint:disable-next-line:ter-arrow-parens
     const ack = async (response: object): Promise<void> => {
       this.logger.debug('calling ack', event.type);
       await this.send(event.envelope_id, response);
```
Comment (connorads) [inline packages/socket-mode/src/SocketModeClient.ts:579]: This was a bit of weird one
image
`tslint` appears to be confused.
image

I tried to "refactor" the implementation by putting the function signature into a `type` to try and trick `tslint` but it didn't work
```typescript
    // Define Ack
    type ackSig = (response: object) => Promise<void>;
    const ack: ackSig = async (response) => {
      this.logger.debug('calling ack', event.type);
      await this.send(event.envelope_id, response);
    };
```

You can't simply remove the arrow syntax as it is needed for the closure & `this`. I had a little play around but gave up and thought can we instead just ignore it? What do you think about adding an ignore for this @seratch?

```typescript
    // Define Ack
    // tslint:disable-next-line
    const ack = async (response: object): Promise<void> => {
      this.logger.debug('calling ack', event.type);
      await this.send(event.envelope_id, response);
    };
```

**Comment by connorads** [comment]
Ran linter locally with Node 12 and Node 16 ✔ Hopefully the rest is fine 👀 Can we run the build please?

File: packages/socket-mode/src/SocketModeClient.ts (hunk 2 above)
Comment (seratch) [reply packages/socket-mode/src/SocketModeClient.ts:579]: >can we instead just ignore it? What do you think about adding an ignore for this @seratch?

Thanks for checking this in details. Ignoring the lint warning sounds fine 👍

**Comment by connorads** [comment]
I took a look at the [Node 16 test failure](https://github.com/slackapi/node-slack-sdk/pull/1249/checks?check_run_id=2722948773) in [`interactive-messages/src/adapter.spec.ts`](https://github.com/slackapi/node-slack-sdk/blob/main/packages/interactive-messages/src/adapter.spec.js#L647-L657).

It doesn't fail on Node 12 locally 🤷‍♂️ (can't see CI because it bails early)
Can recreate it locally with Node 16

It's strange because the test that is reported as failing is not where the failure appears to originate from.
image

And if you run either of those tests, #L647 or #L671, in isolation using `it.only` then they each pass.
Or if you disable the test on #L647 using `xit` then the suite passes - further suggesting the issue originates in that test.

Couldn't seem to get it to pass by changing timeout values or other fiddling.
Any ideas anyone? Anyone got any experience with this test suite/package? @seratch 

Don't know if it's related to #1165

**Comment by seratch** [comment]
@connorads Thanks very much for actively working on this 🙇 

>Couldn't seem to get it to pass by changing timeout values or other fiddling.
>Any ideas anyone? Anyone got any experience with this test suite/package? @seratch
>Don't know if it's related to #1165

Yeah ... I remember the test failure was not reproducible in my environment. It occurred only on the GitHub Actions runtime. Adjusting the timeout again may help.

**Comment by connorads** [comment]
Unfortunately I've hit a bit of a brick wall with this test that only fails with Node 16.
Would someone perhaps be willing to do some remote pairing for 15 mins with me? Or take a look themselves.

I don't think it's necessarily anything to do with timeouts. See the [previous comment](https://github.com/slackapi/node-slack-sdk/pull/1249#issuecomment-852967470) for more context.
Also I think anyone should be able to recreate the issue locally with by checking out this branch and running the tests with Node 16 should someone be able to have a look.

```shell
git clone -b add-node-16-to-ci https://github.com/connorads/node-slack-sdk
cd node-slack-sdk
npm install
npm run setup
npm test
```

**Comment by connorads** [comment]
Had another look at this but haven't made much progress ...

Got some more evidence that it's the test on #L647 causing trouble: If you `.only` run that test then it passes but there's also an error and a stacktrace, which doesn't seem to happen when other tests are run in isolation.
And as said before if you `xit` the test on #L647 then everything passes.

image

I even got desperate and tried updating `mocha` 😅 But that didn't help.

**Comment by connorads** [comment]
Hey @seratch, any chance of getting someone to look at this either with me or by themselves?

_Unfortunately it looks like the person who wrote it @aoberoi no longer works for Slack, but would be great if they could have a look._

**Comment by seratch** [comment]
@connorads As the package is now EOLed, we can disable the flaky test. https://github.com/slackapi/node-slack-sdk/issues/1225 Then, if we don't have any other issues with this PR, we're happy to merge your contribution!

**Review by seratch — APPROVED** [review]

**Comment by seratch** [comment]
@connorads Thanks for your great contribution!

**Pull request merged**